geomesa-cassandra/geomesa-cassandra-datastore/src/main/scala/org/locationtech/geomesa/cassandra/index/CassandraFeatureIndex.scala

```diff
@@ -0,0 +1,186 @@
+/***********************************************************************
+* Copyright (c) 2013-2016 IBM
+* All rights reserved. This program and the accompanying materials
+* are made available under the terms of the Apache License, Version 2.0
+* which accompanies this distribution and is available at
+* http://www.opensource.org/licenses/apache2.0.php.
+*************************************************************************/
+
+package org.locationtech.geomesa.cassandra.index
+
+import com.google.common.primitives.{Bytes, Longs, Ints, Shorts}
+
+import com.datastax.driver.core._
+import com.datastax.driver.core.querybuilder._
+import com.vividsolutions.jts.geom.Geometry
+import org.geotools.factory.Hints
+import org.locationtech.geomesa.cassandra._
+import org.locationtech.geomesa.cassandra.data._
+import org.locationtech.geomesa.curve.BinnedTime
+import org.locationtech.geomesa.index.api.FilterStrategy
+import org.locationtech.geomesa.index.index.ClientSideFiltering.RowAndValue
+import org.locationtech.geomesa.index.index.{ClientSideFiltering, IndexAdapter}
+import org.locationtech.geomesa.curve.TimePeriod
+import org.opengis.feature.simple.SimpleFeatureType
+import org.opengis.feature.simple.SimpleFeature
+import org.opengis.filter.Filter
+
+import scala.collection.JavaConversions._
+
+import java.nio.ByteBuffer
+import java.nio.charset.StandardCharsets
+import java.util.Date
+
+object CassandraFeatureIndex extends CassandraIndexManagerType {
+  // note: keep in priority order for running full table scans
+  override val AllIndices: Seq[CassandraFeatureIndex] =
+    Seq(CassandraZ3Index, CassandraXZ3Index, CassandraZ2Index, CassandraXZ2Index, CassandraIdIndex, CassandraAttributeIndex)
+
+  override val CurrentIndices: Seq[CassandraFeatureIndex] = AllIndices
+}
+
+trait CassandraFeatureIndex extends CassandraFeatureIndexType
+    with IndexAdapter[CassandraDataStore, CassandraFeature, Statement, Statement] with ClientSideFiltering[Row] {
+  import org.locationtech.geomesa.utils.geotools.RichSimpleFeatureType.RichSimpleFeatureType
+
+  type RowToValues = (Array[Byte]) => Any
+
+  var getRowValues:RowToValues = {row => (0L, 0L, "")}
+  var tableName = ""
+  var ks = ""
+
+  def getRowValuesFromSFT(sft: SimpleFeatureType): RowToValues = {
+    // table sharing + shard + 2 byte short + 8 byte long + fid
+    val sharing = sft.getTableSharingPrefix
+    (row) => {
+      var curr = if (sharing == "") 0 else 1
+      // use shard and bin for pk
+      val pk:java.lang.Integer = ((row(curr) & 0xFF) << 16) | ((row(curr + 1) & 0xFF) << 8) | (row(curr + 2) & 0xFF);
+      curr += 3
+      val z = ByteBuffer.wrap(row, curr, 8).getLong
+      curr += 8
+      val fid = new String(row, curr, row.length - curr, StandardCharsets.UTF_8)
+      (pk, z, fid, sharing)
+    }
+  }
+
+  override def configure(sft: SimpleFeatureType, ds: CassandraDataStore): Unit = {
+    super.configure(sft, ds)
+    tableName = getTableName(sft.getTypeName, ds)
```

Comment:
yeah, you don't want any state in this class since it's a globally shared object (the exception would be caches keyed by something unique that's accessible in the method, usually the simple feature type). the idea (which may not work quite right with cassandra) is that the actual values being inserted are separate from the table they are being inserted into. so in our accumulo or hbase impls, we create a mutation here, then in the feature writer we apply it to a particular table. the workflow is like this (using hbase as an example):

1. geomesa feature writer is instantiated and creates 'writers' (details in step 2): https://github.com/locationtech/geomesa/blob/master/geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/geotools/GeoMesaFeatureWriter.scala#L114
2. each index being written to (z3, z2, etc) configures its own writer, which is responsible for creating the insert row key: https://github.com/locationtech/geomesa/blob/master/geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/geotools/GeoMesaFeatureWriter.scala#L95-L97
4. a mutator is created for each index being written to - for hbase this is a buffered mutator: https://github.com/locationtech/geomesa/blob/master/geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/geotools/GeoMesaFeatureWriter.scala#L117
https://github.com/locationtech/geomesa/blob/master/geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/data/HBaseFeatureWriter.scala#L39
5. the mutators are called when a features is actually written after being passed in by the user: https://github.com/locationtech/geomesa/blob/master/geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/geotools/GeoMesaFeatureWriter.scala#L134
https://github.com/locationtech/geomesa/blob/master/geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/data/HBaseFeatureWriter.scala#L43

For C*, you don't really have mutators since everything operates on the session. Possibly when you create the insert row key, you can create a partial statement that has everything except the table defined. Then, when it gets passed to the 'mutators' you could just fill in the table name.

this is probably clear as mud...